### PR TITLE
DOFArray: Pickle/unpickle tags

### PR DESCRIPTION
**Pull request by matthiasdiener**
Potential fix for #317

- [x] Add a test
- [x] Test backwards compatibility

### CODE REVIEW

File: meshmode/dof_array.py [hunk 1]
```diff
@@ -290,7 +292,13 @@ def __setstate__(self, state):
                     "array_context_for_pickling is active.")
 
         self._array_context = actx
-        self._data = tuple([actx.from_numpy(ary_i) for ary_i in state])
+
+        data = state[0]
+        tags = state[1]
```
Comment (inducer) [inline meshmode/dof_array.py:297]:
```suggestion
        data, tags = state
```

File: meshmode/dof_array.py [hunk 2]
```diff
@@ -290,7 +292,13 @@ def __setstate__(self, state):
                     "array_context_for_pickling is active.")
 
         self._array_context = actx
-        self._data = tuple([actx.from_numpy(ary_i) for ary_i in state])
+
+        data = state[0]
```
Comment (inducer) [inline meshmode/dof_array.py:296]: Unless we make this backward-compatible, this will needlessly break existing restart files. (But do include a warning.)

**Review by inducer — COMMENTED** [review]
Thanks! This makes sense. Some comments below.

**Comment by inducer** [comment]
> A test might be necessary

Agree

File: meshmode/dof_array.py [hunk 3]
```diff
@@ -277,7 +277,9 @@ def __getstate__(self):
         if self.array_context is not actx:
             ary = _thaw(actx, _freeze(self))
 
-        return [actx.to_numpy(ary_i) for ary_i in ary._data]
+        return (
+            [actx.to_numpy(ary_i) for ary_i in ary._data],
+            [ary_i.tags for ary_i in ary._data])
```
Comment (kaushikcfd) [inline meshmode/dof_array.py:282]: 1. We should probably store this as a dict?
2. Also needs `.axes` metadata.
3. What about some array contexts that operate on arrays that do not have `.tags` attribute?
4. Should we care about backward compatibility with the existing data that was pickled before this PR?

**Review by kaushikcfd — CHANGES_REQUESTED** [review]
Thanks, left some questions to discuss. This also needs some tests.

File: meshmode/dof_array.py (hunk 1 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py]: Updated by [cfa2262](https://github.com/inducer/meshmode/pull/318/commits/cfa226289c43c89b278e03a988aaad8f8599a2ad)

File: meshmode/dof_array.py (hunk 2 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py]: I *think* [cfa2262](https://github.com/inducer/meshmode/pull/318/commits/cfa226289c43c89b278e03a988aaad8f8599a2ad), but haven't tested yet.

File: meshmode/dof_array.py (hunk 3 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py]: > 1. We should probably store this as a dict?
Done in https://github.com/inducer/meshmode/pull/318/commits/2283819975aa5b0fdbc38e1abaf987aafc6e84e8
> 2. Also needs `.axes` metadata.
Done in https://github.com/inducer/meshmode/pull/318/commits/de6e398d5d8b4605a3a3dc5b8d870be3e514d418
> 3. What about some array contexts that operate on arrays that do not have `.tags` attribute?
https://github.com/inducer/meshmode/pull/318/commits/17e63d660cb4e0ccb23e9ecb4b3242aa108efcbb should hopefully fix that
> 4. Should we care about backward compatibility with the existing data that was pickled before this PR?

See [cfa2262](https://github.com/inducer/meshmode/pull/318/commits/cfa226289c43c89b278e03a988aaad8f8599a2ad)

**Comment by matthiasdiener** [comment]
This is ready for a second look. There are some strange CI errors that are probably unrelated to this PR.

**Comment by inducer** [comment]
> There are some strange CI errors that are probably unrelated to this PR.

Those were my fault. I screwed something up in ci-support.

File: meshmode/dof_array.py [hunk 4]
```diff
@@ -277,7 +277,16 @@ def __getstate__(self):
...
 
-        return [actx.to_numpy(ary_i) for ary_i in ary._data]
+        d = {}
+        d["data"] = [actx.to_numpy(ary_i) for ary_i in ary._data]
+
+        if hasattr(ary._data[0], "tags"):
+            d["tags"] = [ary_i.tags for ary_i in ary._data]
+        else:
+            d["tags"] = [frozenset() for _ in range(len(ary._data))]
+        d["axes_tags"] = [ax.tags for ary_i in ary._data for ax in ary_i.axes]
```
Comment (kaushikcfd) [inline meshmode/dof_array.py:287]: 1. Another `hasattr(..., "axes")` here?
2. A more natural way to store it is a list of list? The inner list stores the tags corresponding to each axis of the array.

**Review by kaushikcfd — APPROVED** [review]
Thanks! Suggested some minor changes, otherwise LGTM.

File: meshmode/dof_array.py (hunk 4 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py]: Is 
89d7c19 what you had in mind?

File: meshmode/dof_array.py [hunk 5]
```diff
@@ -277,7 +277,20 @@ def __getstate__(self):
...
+
+        if hasattr(ary._data[0], "tags"):
+            d["tags"] = [ary_i.tags for ary_i in ary._data]
+        else:
+            d["tags"] = [frozenset() for _ in range(len(ary._data))]
+
+        if hasattr(ary._data[0], "axes"):
+            d["axes_tags"] = [[ax.tags for ax in ary_i.axes] for ary_i in ary._data]
+        else:
+            d["axes_tags"] = [[] for _ in range(len(ary._data))]
```
Comment (kaushikcfd) [inline meshmode/dof_array.py:291]: This should be: `d["axes_tags"] = [[frozenset() for _ in range(leaf_ary.ndim)] for leaf_ary in range(len(ary._data))]`

File: meshmode/dof_array.py [hunk 6]
```diff
@@ -290,7 +303,28 @@ def __setstate__(self, state):
...
+
+        if isinstance(state, dict):
+            data = state["data"]
+            tags = state["tags"]
+            axes_tags = state["axes_tags"]
+        else:
+            # For backwards compatibility
+            data = state
+            tags = [frozenset() for _ in range(len(state))]
+            axes_tags = [frozenset() for _ in range(len(state))]
```
Comment (kaushikcfd) [inline meshmode/dof_array.py:315]: This default is incorrect, according to the implementation in `__getstate__`.

File: meshmode/dof_array.py [hunk 7]
```diff
@@ -290,7 +303,28 @@ def __setstate__(self, state):
...
+
+        self._data = []
+
```
Comment (kaushikcfd) [inline meshmode/dof_array.py:318]: How about an assert here that `assert len(data) == len(tags) == len(axes_tags)`?

File: meshmode/dof_array.py [hunk 8]
```diff
@@ -290,7 +303,28 @@ def __setstate__(self, state):
...
+        for idx, ary in enumerate(data):
+            d = actx.tag(tags[idx], actx.from_numpy(ary))
+
```
Comment (kaushikcfd) [inline meshmode/dof_array.py:321]: How about an assert here that `assert len(axes_tags[idx]) == ary.ndim`?

File: meshmode/dof_array.py [hunk 9]
```diff
@@ -290,7 +303,28 @@ def __setstate__(self, state):
...
+            for ida, ax in enumerate(axes_tags[idx]):
+                d = actx.tag_axis(ida, ax, d)
```
Comment (kaushikcfd) [inline meshmode/dof_array.py:323]: How about an assert before this line that `assert isinstance(axes_tags[idx], frozenset)`?

File: meshmode/dof_array.py [hunk 10]
```diff
@@ -277,7 +277,20 @@ def __getstate__(self):
...
             ary = _thaw(actx, _freeze(self))
 
-        return [actx.to_numpy(ary_i) for ary_i in ary._data]
+        d = {}
+        d["data"] = [actx.to_numpy(ary_i) for ary_i in ary._data]
+
+        if hasattr(ary._data[0], "tags"):
+            d["tags"] = [ary_i.tags for ary_i in ary._data]
+        else:
+            d["tags"] = [frozenset() for _ in range(len(ary._data))]
```
Comment (kaushikcfd) [inline meshmode/dof_array.py:286]: Is it reasonable to expect that `ary._data` has at least 1 array? (I'm not sure, someone else can comment on this?)

**Review by kaushikcfd — APPROVED** [review]
Thanks. Found some minor bugs, and suggested some asserted that would (I think) catch such errors sooner.

File: meshmode/dof_array.py (hunk 7 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py:322]: Added in bd439cc159d69b0647375fb4c71f2d3734c227ee, thanks!

File: meshmode/dof_array.py (hunk 8 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py:327]: Added in bd439cc159d69b0647375fb4c71f2d3734c227ee, thanks!

File: meshmode/dof_array.py (hunk 9 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py:329]: `axes_tags[idx]` is a list; added this assert in bd439cc159d69b0647375fb4c71f2d3734c227ee

File: meshmode/dof_array.py (hunk 5 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py]: I changed it to `d["axes_tags"] = [[frozenset() for _ in range(leaf_ary.ndim)]
                                         for leaf_ary in ary._data]`

File: meshmode/dof_array.py (hunk 6 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py]: Thanks, I changed it to `axes_tags = [[frozenset() for _ in range(leaf_ary.ndim)]
                                         for leaf_ary in data]`

File: meshmode/dof_array.py (hunk 10 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py]: It seems an empty DOFArray is legal, so I added support for this case in 1ae79bd

File: meshmode/dof_array.py [hunk 11]
```diff
@@ -277,7 +277,21 @@ def __getstate__(self):
...
             ary = _thaw(actx, _freeze(self))
 
-        return [actx.to_numpy(ary_i) for ary_i in ary._data]
+        d = {}
+        d["data"] = [actx.to_numpy(ary_i) for ary_i in ary._data]
+
+        if len(ary) > 0 and hasattr(ary._data[0], "tags"):
+            d["tags"] = [ary_i.tags for ary_i in ary._data]
+        else:
+            d["tags"] = [frozenset() for _ in range(len(ary._data))]
```
Comment (alexfikl) [inline meshmode/dof_array.py:286]:
```suggestion
	d["tags"] = [getattr(ary_i, "tags", frozenset()) for ary_i in ary]
```
This looks more readable to me and it should catch all the cases. What do you think?

Comment (alexfikl) [reply meshmode/dof_array.py:286]: There's probably a similar rewrite for the `axes`, but that's not going to look as nice because of the nesting :\

Comment (matthiasdiener) [reply meshmode/dof_array.py]: Looks good, thanks!

File: meshmode/dof_array.py [hunk 12]
```diff
@@ -290,7 +301,33 @@ def __setstate__(self, state):
...
+        if isinstance(state, dict):
+            data = state["data"]
+            tags = state["tags"]
+            axes_tags = state["axes_tags"]
+        else:
+            # For backwards compatibility
+            data = state
+            tags = [frozenset() for _ in range(len(data))]
+            axes_tags = [[frozenset() for _ in range(leaf_ary.ndim)]
+                                         for leaf_ary in data]
```
Comment (inducer) [inline meshmode/dof_array.py:314]: Please add a warning here that a `DOFArray` is being unpickled without metadata, and that program transformation may fail as a result.

**Comment by kaushikcfd** [comment]
The point at which metadata is being recorded may not have everything gone through metadata inference. Needs a freeze, something like this:

```diff
[line@line meshmode]$ git diff
diff --git a/meshmode/dof_array.py b/meshmode/dof_array.py
index eb3cd3f..4375228 100644
--- a/meshmode/dof_array.py
+++ b/meshmode/dof_array.py
@@ -272,7 +272,7 @@ class DOFArray:
             raise RuntimeError("DOFArray instances can only be pickled while "
                     "array_context_for_pickling is active.")
 
-        ary = self
+        ary = _thaw(freeze(self, self.array_context), self.array_context)
 
         if self.array_context is not actx:
             ary = _thaw(actx, _freeze(self))
@@ -286,7 +286,7 @@ class DOFArray:
             d["axes_tags"] = [[ax.tags for ax in ary_i.axes] for ary_i in ary._data]
         else:
             d["axes_tags"] = [[frozenset() for _ in range(leaf_ary.ndim)]
-                                         for leaf_ary in ary._data]
+                              for leaf_ary in ary._data]
 
         return d
```

(We have something similar in arraycontext-land: https://github.com/inducer/arraycontext/pull/152)

**Comment by inducer** [comment]
> Needs a freeze, something like this:

Also add a comment justifying the thaw/freeze, maybe including a link back to Kaushik's comment.

File: meshmode/dof_array.py (hunk 12 above)
Comment (matthiasdiener) [reply meshmode/dof_array.py:319]: Added in 
137e4dc

**Comment by matthiasdiener** [comment]
> The point at which metadata is being recorded may not have everything gone through metadata inference. Needs a freeze, something like this:
> 

Added in [0438b18](https://github.com/inducer/meshmode/pull/318/commits/0438b18dc4ac7746f08d72ca1ca740a96398472c)

File: test/test_array.py [hunk 13]
```diff
@@ -177,6 +177,43 @@ def test_dof_array_pickling(actx_factory):
     assert flat_norm(mat_of_dofs - mat2_of_dofs, np.inf) == 0
     assert flat_norm(dc_of_dofs - dc2_of_dofs, np.inf) == 0
 
+
+def test_dof_array_pickling_tags(actx_factory):
```
Comment (inducer) [inline test/test_array.py:181]: I don't think this test is being run against a lazy array context. I think that would be important to do.

Comment (matthiasdiener) [reply test/test_array.py]: I added testing with pytato to this file in [c82c6c2](https://github.com/inducer/meshmode/pull/318/commits/c82c6c251c36eb31e50839a7e1428814c1439946)

**Comment by inducer** [comment]
LGTM, thanks!

**Pull request merged**